### ISSUE TITLE
unused_declaration does not catch View only used in PreviewPeovider

### ISSUE THREAD

**Issue by iressler** [realm/SwiftLint#4720]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

When a View is only used by it's matching PreviewProvider `unused_declarations` does not report it as unused. E.g. if this is in a file and the View is not referenced anywhere else SwiftLint will not report any issues:

```Swift
struct UnusedView: View {
  var body: some View {
    Text("Hello, World!")
  }
}

struct UnusedView_Previews: PreviewProvider {
  static var previews: some View {
    UnusedView()
  }
}
```

This may be expected behavior, the View is being used, however I was surprised by it and don't consider that a "real" use. If this isn't reported as unused then no SwiftUI View with a Preview will ever be "unused", even though they aren't actually being used anywhere. 

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ swiftlint analyze --compiler-log-path <path>
Analyzing Swift files in current working directory
Collecting 'UnusedDeclarationsPreviewsApp.swift' (1/3)
Collecting 'ContentView.swift' (2/3)
Collecting 'UnusedView.swift' (3/3)
Analyzing 'UnusedDeclarationsPreviewsApp.swift' (1/3)
Analyzing 'ContentView.swift' (2/3)
Analyzing 'UnusedView.swift' (3/3)
Done analyzing! Found 0 violations, 0 serious in 3 files.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)?
  0.50.3
* Installation method used (Homebrew, CocoaPods, building from source, etc)?
  HomeBrew
* Paste your configuration file:

```yml
analyzer_rules:
  - unused_declaration
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)?
  No
* Which Xcode version are you using (check `xcodebuild -version`)?
  Xcode 14.2
  Build version 14C18
* Do you have a sample that shows the issue?
  See above for code, but I've also provided a [sample project](https://github.com/realm/SwiftLint/files/10492675/UnusedDeclarationsPreviews.zip).

**Comment by SimplyDanny** [comment]
Seems to be a good idea to add an option to the rule that makes it treat types only used in previews as unused.
